core/src/main/java/org/apache/brooklyn/core/config/internal/LazyContainerAndKeyValue.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.config.internal;
+
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.config.ConfigValueAtContainer;
+import org.apache.brooklyn.util.guava.Maybe;
+
+import com.google.common.base.Function;
+
+public class LazyContainerAndKeyValue<TContainer,TValue> implements ConfigValueAtContainer<TContainer,TValue> {
+    
+    private final TContainer container;
+    private final ConfigKey<TValue> key;
+    private final Function<TContainer,Maybe<Object>> lookupResolutionFunction;
+    private final Function<Maybe<Object>,Maybe<TValue>> conversionFunction;
+    private Maybe<TValue> resolved;
+    
+    public LazyContainerAndKeyValue(ConfigKey<TValue> key, TContainer container, 
+            Function<TContainer, Maybe<Object>> lookupResolutionFunction,
+            Function<Maybe<Object>, Maybe<TValue>> conversionFunction) {
+        this.key = key;
```

Comment:
We should use `this.key = checkNotNull(key, "key")` (or if it really is allowed to be null then use `@Nullable` on parameter, and on `getKey` method.

Same for the other args.